File: src/main/scala/org/apache/spark/metrics/reporter/LocalCsvReporter.java
package org.apache.spark.metrics.reporter;

import com.codahale.metrics.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.*;
import java.util.Locale;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

import static java.nio.charset.StandardCharsets.UTF_8;

/**
 * A reporter which creates a comma-separated values file of the measurements for each metric.
 */
public class LocalCsvReporter extends AbstractCsvReporter {

    /**
     * A builder for {@link LocalCsvReporter} instances. Defaults to using the default locale, converting
     * rates to events/second, converting durations to milliseconds, and not filtering metrics.
     */

    private static final Logger LOGGER = LoggerFactory.getLogger(LocalCsvReporter.class);

    private final CsvFileProvider csvFileProvider;
    private final File directoryFile;

    protected LocalCsvReporter(String directory,
                               MetricRegistry registry,
                               Locale locale,
                               String separator,
                               TimeUnit rateUnit,
                               TimeUnit durationUnit,
                               Clock clock,
                               MetricFilter filter,
                               ScheduledExecutorService executor,
                               boolean shutdownExecutorOnStop,
                               CsvFileProvider csvFileProvider) {
        super(registry, locale, directory, separator, rateUnit, durationUnit, clock, filter, executor, shutdownExecutorOnStop);
        LOGGER.info("Using LocalCsvReporter");
        this.csvFileProvider = csvFileProvider;
        this.directoryFile = new File(directory.replace("file:", ""));
    }

    protected void report(long timestamp, String name, String header, String line, Object... values) {
        try {
            final File file = csvFileProvider.getFile(directoryFile, name);
            final boolean fileAlreadyExists = file.exists();
            if (fileAlreadyExists || file.createNewFile()) {
                try (PrintWriter out = new PrintWriter(new OutputStreamWriter(
                        new FileOutputStream(file, true), UTF_8))) {
                    if (!fileAlreadyExists) {
                        out.println("t" + separator + header);
                    }
                    out.printf(locale, String.format(locale, "%d" + separator + "%s%n", timestamp, line), values);
                }
            }
        } catch (IOException e) {
            LOGGER.warn("Error writing {} to local dir {}", name, directory, e);
        }
    }
}

File: src/main/scala/org/apache/spark/metrics/reporter/LocalCsvReporter.scala
package org.apache.spark.metrics.reporter

import com.codahale.metrics._
import com.ucesys.sparkscope.common.SparkScopeLogger
import com.ucesys.sparkscope.data.DataTable
import com.ucesys.sparkscope.io.writer.LocalFileWriter

import java.io.IOException
import java.util.Locale
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import java.nio.file.Paths

/**
 * A reporter which creates a comma-separated values file of the measurements for each metric.
 */
class LocalCsvReporter(rootDir: String,
                       registry: MetricRegistry,
                       locale: Locale,
                       separator: String,
                       rateUnit: TimeUnit,
                       durationUnit: TimeUnit,
                       clock: Clock,
                       filter: MetricFilter,
                       fileWriter: LocalFileWriter,
                       appName: Option[String] = None)
                      (implicit logger: SparkScopeLogger)
  extends AbstractCsvReporter(registry, locale, separator, rateUnit, durationUnit, clock, filter) {

    logger.info("Using LocalCsvReporter", this.getClass)

    override protected[reporter] def report(appId: String, instance: String, metrics: DataTable, timestamp: Long): Unit = {
        logger.debug("\n" + metrics.toString, this.getClass)
        val row: String = metrics.toCsvNoHeader(separator)

        val appDir = Paths.get(rootDir, appName.getOrElse(""), appId).toString
        val csvFilePath = Paths.get(appDir, s"${instance}.csv").toString

        try {
            logger.debug(s"Writing to ${csvFilePath}", this.getClass)
            if (!fileWriter.exists(csvFilePath)) {
                fileWriter.makeDir(appDir)
                fileWriter.write(csvFilePath, metrics.header + "\n");
            }

            fileWriter.append(csvFilePath, row);
        } catch {
            case e: IOException => logger.warn(s"Error writing ${metrics.name} to local dir ${csvFilePath}. ${e}", this.getClass)
        }
    }

}
